[CI] SharedBlobCacheWarmingServiceIT testCacheIsWarmedBeforeSearchShardRecovery failing

**Build Scans:**
- [elasticsearch-serverless-validate-submodule #22526 / java_run_serverless_elasticsearch_tests](https://gradle-enterprise.elastic.co/s/2gffiylcxvsai)
- [elasticsearch-serverless-es-pr-check #73792 / java_checkpart1](https://gradle-enterprise.elastic.co/s/rn66lnjgwgwek)
- [elasticsearch-serverless-es-pr-check #73774 / java_checkpart1](https://gradle-enterprise.elastic.co/s/36tixekbxk4ds)
- [elasticsearch-serverless-es-pr-check #73769 / java_checkpart1](https://gradle-enterprise.elastic.co/s/e2kzsa5lp4rpu)
- [elasticsearch-serverless-es-pr-check #73745 / java_checkpart1](https://gradle-enterprise.elastic.co/s/vlnosokirafbo)
- [elasticsearch-serverless-es-pr-check #73754 / java_checkpart1](https://gradle-enterprise.elastic.co/s/3eflnrzkweesa)
- [elasticsearch-serverless-es-pr-check #73695 / java_checkpart1](https://gradle-enterprise.elastic.co/s/uczff7kcfblxo)
- [elasticsearch-serverless-es-pr-check #73681 / java_checkpart1](https://gradle-enterprise.elastic.co/s/4g274b2dk6eru)
- [elasticsearch-serverless-es-pr-check #73672 / java_checkpart1](https://gradle-enterprise.elastic.co/s/ye3hcw4wia7iq)
- [elasticsearch-serverless-es-pr-check #73630 / java_checkpart1](https://gradle-enterprise.elastic.co/s/eku57xpduiuio)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:internalClusterTest" --tests "org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT.testCacheIsWarmedBeforeSearchShardRecovery" -Dtests.seed=2F5C0A130896A50C -Dtests.locale=bez -Dtests.timezone=Asia/Calcutta -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testCacheIsWarmedBeforeSearchShardRecovery'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: timed out waiting for green state
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testCacheIsWarmedBeforeSearchShardRecovery
- [main] 1 consecutive failures in step java_run_serverless_elasticsearch_tests
- [main] 1 consecutive failures in pipeline elasticsearch-serverless-validate-submodule
- [main] 11 failures in test testCacheIsWarmedBeforeSearchShardRecovery (3.6% fail rate in 308 executions)
- [main] 10 failures in step java_checkpart1 (4.2% fail rate in 240 executions)
- [main] 10 failures in pipeline elasticsearch-serverless-es-pr-check (4.5% fail rate in 222 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

---

Migrated from https://github.com/elastic/elasticsearch-serverless/issues/6096

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -406,9 +406,6 @@ tests:
 - class: "org.elasticsearch.xpack.stateless.reshard.StatelessReshardMixedOperationsIT"
   issue: "https://github.com/elastic/elasticsearch/issues/147842"
   method: "testMixedOperationsDuringSplitWithDisruption"
-- class: org.elasticsearch.xpack.stateless.cache.SharedBlobCacheWarmingServiceIT
-  method: testCacheIsWarmedBeforeSearchShardRecovery
-  issue: https://github.com/elastic/elasticsearch/issues/147843
 - class: org.elasticsearch.xpack.stateless.StatelessHollowIndexShardsIT
   method: testStressWithRelocationOrObjectStoreFailures
   issue: https://github.com/elastic/elasticsearch/issues/147844

--- a/x-pack/plugin/stateless/src/main/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingService.java
+++ b/x-pack/plugin/stateless/src/main/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingService.java
@@ -98,7 +98,9 @@ public class SharedBlobCacheWarmingService {
         INDEXING_EARLY(true),
         INDEXING(true),
         INDEXING_MERGE(false),
-        SEARCH(true),
+        // search shard recovery doesn't guarantee that all of region 0 has been cached, because header reads served from
+        // index shards are served at page rather than region granularity.
+        SEARCH(false),
         HOLLOWING(true),
         UNHOLLOWING(true);
 
@@ -1045,22 +1047,22 @@ public class SharedBlobCacheWarmingService {
         }
 
         private void addLocation(BlobLocation location, String fileName, long position, long length, ActionListener<Void> listener) {
-            final long start = position;
-            final long end = position + length;
+            final long start = length <= Integer.MAX_VALUE ? directory.getPosition(fileName, position, (int) length) : position;
+            final long end = start + length;
             final int regionSize = cacheService.getRegionSize();
             final int startRegion = cacheService.getRegion(start);
             final int endRegion = cacheService.getEndingRegion(end);
 
             if (startRegion == endRegion) {
                 BlobRegion blobRegion = new BlobRegion(location.blobFile(), startRegion);
-                enqueueLocation(blobRegion, fileName, location, position, length, listener);
+                enqueueLocation(blobRegion, location, start, length, listener);
             } else {
                 try (var listeners = new RefCountingListener(listener)) {
                     for (int r = startRegion; r <= endRegion; r++) {
                         // adjust the position & length to the region
                         var range = ByteRange.of(Math.max(start, (long) r * regionSize), Math.min(end, (r + 1L) * regionSize));
                         BlobRegion blobRegion = new BlobRegion(location.blobFile(), r);
-                        enqueueLocation(blobRegion, fileName, location, range.start(), range.length(), listeners.acquire());
+                        enqueueLocation(blobRegion, location, range.start(), range.length(), listeners.acquire());
                     }
                 }
             }
@@ -1108,29 +1110,25 @@ public class SharedBlobCacheWarmingService {
             }));
         }
 
-        private boolean shouldSkipLocationWarming(String fileName, long position, long length) {
+        private boolean shouldSkipLocationWarming(long position) {
             if (warmingRun.type.skipsWarmingForRegion0Locations == false) {
                 return false;
             }
-            if (length > Short.MAX_VALUE) {
-                // length is too long to be contained in replicated section
-                return false;
-            }
-            int region = (int) (directory.getPosition(fileName, position, (int) length) / cacheService.getRegionSize());
             // region 0 is already loaded by this point while resolving full set of commit files and safe to skip.
             // See org.elasticsearch.xpack.stateless.objectstore.ObjectStoreService#readIndexingShardState
-            return region == 0;
+            // reads that span regions will be split into separate tasks, so checking the starting position suffices.
+            return position / cacheService.getRegionSize() == 0;
         }
 
+        // When using replicated ranges, the location should already be translated to its replicated counterpart
         private void enqueueLocation(
             BlobRegion blobRegion,
-            String fileName,
             BlobLocation blobLocation,
             long position,
             long length,
             ActionListener<Void> listener
         ) {
-            if (shouldSkipLocationWarming(fileName, position, length)) {
+            if (shouldSkipLocationWarming(position)) {
                 skippedTasksCount.incrementAndGet();
                 listener.onResponse(null);
                 return;

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceTests.java
@@ -895,9 +895,8 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
             node.indexingDirectory.getBlobStoreCacheDirectory().updateMetadata(blobFileRanges, randomNonNegativeLong());
 
             final PlainActionFuture<Void> future = new PlainActionFuture<>();
-            var type = randomFrom(INDEXING, SEARCH);
             node.warmingService.warmCache(
-                type,
+                INDEXING,
                 mockIndexShard(node),
                 commit,
                 node.indexingDirectory.getBlobStoreCacheDirectory(),

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceTests.java
@@ -29,7 +29,9 @@ import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.common.util.concurrent.EsExecutors;
 import org.elasticsearch.core.Releasable;
+import org.elasticsearch.env.Environment;
 import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.env.TestEnvironment;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.IndexShardState;
 import org.elasticsearch.index.shard.ShardId;
@@ -89,12 +91,10 @@ import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
 import static org.hamcrest.Matchers.lessThanOrEqualTo;
-import static org.mockito.AdditionalMatchers.not;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -852,21 +852,31 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
     }
 
     public void testOnlyTheFirstRegionIsLoadedWhenReplicatedContentIsPresent() throws IOException {
+        var primaryTerm = 1;
         var cacheSize = ByteSizeValue.ofMb(10L);
         var regionSize = ByteSizeValue.ofBytes((long) randomIntBetween(1, 3) * 2 * SharedBytes.PAGE_SIZE);
-        try (var node = new FakeStatelessNodeWithFakeObjectStore() {
-            @Override
-            protected Settings nodeSettings() {
-                return Settings.builder()
-                    .put(super.nodeSettings())
-                    .put(SharedBlobCacheService.SHARED_CACHE_SIZE_SETTING.getKey(), cacheSize)
-                    .put(SharedBlobCacheService.SHARED_CACHE_REGION_SIZE_SETTING.getKey(), regionSize)
-                    .put(SharedBlobCacheService.SHARED_CACHE_RANGE_SIZE_SETTING.getKey(), regionSize)
-                    .put(SharedBlobCacheWarmingService.PREWARMING_RANGE_MINIMIZATION_STEP.getKey(), regionSize)
-                    .put(StatelessCommitService.STATELESS_COMMIT_USE_INTERNAL_FILES_REPLICATED_CONTENT.getKey(), true)
-                    .build();
+        try (
+            var node = new FakeStatelessNode(
+                TestEnvironment::newEnvironment,
+                settings -> new NodeEnvironment(settings, TestEnvironment.newEnvironment(settings)),
+                xContentRegistry(),
+                primaryTerm
+            ) {
+                @Override
+                protected Settings nodeSettings() {
+                    return Settings.builder()
+                        .put(super.nodeSettings())
+                        .put(Environment.PATH_HOME_SETTING.getKey(), createTempDir())
+                        .putList(Environment.PATH_DATA_SETTING.getKey(), createTempDir().toAbsolutePath().toString())
+                        .put(SharedBlobCacheService.SHARED_CACHE_SIZE_SETTING.getKey(), cacheSize)
+                        .put(SharedBlobCacheService.SHARED_CACHE_REGION_SIZE_SETTING.getKey(), regionSize)
+                        .put(SharedBlobCacheService.SHARED_CACHE_RANGE_SIZE_SETTING.getKey(), regionSize)
+                        .put(SharedBlobCacheWarmingService.PREWARMING_RANGE_MINIMIZATION_STEP.getKey(), regionSize)
+                        .put(StatelessCommitService.STATELESS_COMMIT_USE_INTERNAL_FILES_REPLICATED_CONTENT.getKey(), true)
+                        .build();
+                }
             }
-        }) {
+        ) {
             final var primaryTermAndGeneration = new PrimaryTermAndGeneration(randomNonNegativeLong(), randomLongBetween(3, 42));
             final var blobFile = new BlobFile(
                 StatelessCompoundCommit.blobNameFromGeneration(primaryTermAndGeneration.generation()),
@@ -896,15 +906,10 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
                 future
             );
             safeGet(future);
-            verify(node.sharedCacheService, never()).maybeFetchRange(
-                any(),
-                not(eq(0)),
-                any(),
-                anyLong(),
-                any(),
-                any(),
-                anyActionListener()
-            );
+
+            SharedBlobCacheService.Stats stats = node.sharedCacheService.getStats();
+            assertThat(stats.writeCount(), equalTo(0L));
+            assertThat(stats.missCount(), equalTo(0L));
         }
     }
 
@@ -1170,7 +1175,7 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
         boolean preWarmForIdLookupEnabled,
         double idLookupPreWarmRatio
     ) throws IOException {
-        return new FakeStatelessNodeWithFakeObjectStore() {
+        return new FakeStatelessNode(this::newEnvironment, this::newNodeEnvironment, xContentRegistry()) {
             @Override
             protected Settings nodeSettings() {
                 Settings settings = super.nodeSettings();
@@ -1187,6 +1192,43 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
                     .put(SharedBlobCacheWarmingService.PREWARM_INDEX_SHARD_FOR_ID_LOOKUPS_SETTING.getKey(), preWarmForIdLookupEnabled)
                     .put(SharedBlobCacheWarmingService.ID_LOOKUP_PREWARM_RATIO_SETTING.getKey(), idLookupPreWarmRatio)
                     .build();
+            }
+
+            @Override
+            protected StatelessSharedBlobCacheService createCacheService(
+                NodeEnvironment nodeEnvironment,
+                Settings settings,
+                ThreadPool threadPool,
+                MeterRegistry meterRegistry
+            ) {
+                return spy(super.createCacheService(nodeEnvironment, settings, threadPool, meterRegistry));
+            }
+
+            @Override
+            public BlobContainer wrapBlobContainer(BlobPath path, BlobContainer innerContainer) {
+                return new FilterBlobContainer(innerContainer) {
+                    @Override
+                    protected BlobContainer wrapChild(BlobContainer child) {
+                        return child;
+                    }
+
+                    @Override
+                    public InputStream readBlob(OperationPurpose purpose, String blobName, long position, long length) throws IOException {
+                        return new InputStream() {
+                            private long remaining = length;
+
+                            @Override
+                            public int read() throws IOException {
+                                if (remaining == 0) {
+                                    return -1;
+                                } else {
+                                    remaining -= 1;
+                                    return 1;
+                                }
+                            }
+                        };
+                    }
+                };
             }
         };
     }
@@ -1229,58 +1271,6 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
                     .build();
             }
         };
-    }
-
-    class FakeStatelessNodeWithFakeObjectStore extends FakeStatelessNode {
-        FakeStatelessNodeWithFakeObjectStore() throws IOException {
-            this(1);
-        }
-
-        FakeStatelessNodeWithFakeObjectStore(int primaryTerm) throws IOException {
-            super(
-                SharedBlobCacheWarmingServiceTests.this::newEnvironment,
-                SharedBlobCacheWarmingServiceTests.this::newNodeEnvironment,
-                xContentRegistry(),
-                primaryTerm
-            );
-        }
-
-        @Override
-        protected StatelessSharedBlobCacheService createCacheService(
-            NodeEnvironment nodeEnvironment,
-            Settings settings,
-            ThreadPool threadPool,
-            MeterRegistry meterRegistry
-        ) {
-            return spy(super.createCacheService(nodeEnvironment, settings, threadPool, meterRegistry));
-        }
-
-        @Override
-        public BlobContainer wrapBlobContainer(BlobPath path, BlobContainer innerContainer) {
-            return new FilterBlobContainer(innerContainer) {
-                @Override
-                protected BlobContainer wrapChild(BlobContainer child) {
-                    return child;
-                }
-
-                @Override
-                public InputStream readBlob(OperationPurpose purpose, String blobName, long position, long length) throws IOException {
-                    return new InputStream() {
-                        private long remaining = length;
-
-                        @Override
-                        public int read() throws IOException {
-                            if (remaining == 0) {
-                                return -1;
-                            } else {
-                                remaining -= 1;
-                                return 1;
-                            }
-                        }
-                    };
-                }
-            };
-        }
     }
 
     private void appendCommitsToVbcc(

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cache/SharedBlobCacheWarmingServiceTests.java
@@ -29,9 +29,7 @@ import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.common.util.concurrent.EsExecutors;
 import org.elasticsearch.core.Releasable;
-import org.elasticsearch.env.Environment;
 import org.elasticsearch.env.NodeEnvironment;
-import org.elasticsearch.env.TestEnvironment;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.IndexShardState;
 import org.elasticsearch.index.shard.ShardId;
@@ -91,10 +89,12 @@ import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
 import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.mockito.AdditionalMatchers.not;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -852,31 +852,21 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
     }
 
     public void testOnlyTheFirstRegionIsLoadedWhenReplicatedContentIsPresent() throws IOException {
-        var primaryTerm = 1;
         var cacheSize = ByteSizeValue.ofMb(10L);
         var regionSize = ByteSizeValue.ofBytes((long) randomIntBetween(1, 3) * 2 * SharedBytes.PAGE_SIZE);
-        try (
-            var node = new FakeStatelessNode(
-                TestEnvironment::newEnvironment,
-                settings -> new NodeEnvironment(settings, TestEnvironment.newEnvironment(settings)),
-                xContentRegistry(),
-                primaryTerm
-            ) {
-                @Override
-                protected Settings nodeSettings() {
-                    return Settings.builder()
-                        .put(super.nodeSettings())
-                        .put(Environment.PATH_HOME_SETTING.getKey(), createTempDir())
-                        .putList(Environment.PATH_DATA_SETTING.getKey(), createTempDir().toAbsolutePath().toString())
-                        .put(SharedBlobCacheService.SHARED_CACHE_SIZE_SETTING.getKey(), cacheSize)
-                        .put(SharedBlobCacheService.SHARED_CACHE_REGION_SIZE_SETTING.getKey(), regionSize)
-                        .put(SharedBlobCacheService.SHARED_CACHE_RANGE_SIZE_SETTING.getKey(), regionSize)
-                        .put(SharedBlobCacheWarmingService.PREWARMING_RANGE_MINIMIZATION_STEP.getKey(), regionSize)
-                        .put(StatelessCommitService.STATELESS_COMMIT_USE_INTERNAL_FILES_REPLICATED_CONTENT.getKey(), true)
-                        .build();
-                }
+        try (var node = new FakeStatelessNodeWithFakeObjectStore() {
+            @Override
+            protected Settings nodeSettings() {
+                return Settings.builder()
+                    .put(super.nodeSettings())
+                    .put(SharedBlobCacheService.SHARED_CACHE_SIZE_SETTING.getKey(), cacheSize)
+                    .put(SharedBlobCacheService.SHARED_CACHE_REGION_SIZE_SETTING.getKey(), regionSize)
+                    .put(SharedBlobCacheService.SHARED_CACHE_RANGE_SIZE_SETTING.getKey(), regionSize)
+                    .put(SharedBlobCacheWarmingService.PREWARMING_RANGE_MINIMIZATION_STEP.getKey(), regionSize)
+                    .put(StatelessCommitService.STATELESS_COMMIT_USE_INTERNAL_FILES_REPLICATED_CONTENT.getKey(), true)
+                    .build();
             }
-        ) {
+        }) {
             final var primaryTermAndGeneration = new PrimaryTermAndGeneration(randomNonNegativeLong(), randomLongBetween(3, 42));
             final var blobFile = new BlobFile(
                 StatelessCompoundCommit.blobNameFromGeneration(primaryTermAndGeneration.generation()),
@@ -906,10 +896,15 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
                 future
             );
             safeGet(future);
-
-            SharedBlobCacheService.Stats stats = node.sharedCacheService.getStats();
-            assertThat(stats.writeCount(), equalTo(0L));
-            assertThat(stats.missCount(), equalTo(0L));
+            verify(node.sharedCacheService, never()).maybeFetchRange(
+                any(),
+                not(eq(0)),
+                any(),
+                anyLong(),
+                any(),
+                any(),
+                anyActionListener()
+            );
         }
     }
 
@@ -1175,7 +1170,7 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
         boolean preWarmForIdLookupEnabled,
         double idLookupPreWarmRatio
     ) throws IOException {
-        return new FakeStatelessNode(this::newEnvironment, this::newNodeEnvironment, xContentRegistry()) {
+        return new FakeStatelessNodeWithFakeObjectStore() {
             @Override
             protected Settings nodeSettings() {
                 Settings settings = super.nodeSettings();
@@ -1192,43 +1187,6 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
                     .put(SharedBlobCacheWarmingService.PREWARM_INDEX_SHARD_FOR_ID_LOOKUPS_SETTING.getKey(), preWarmForIdLookupEnabled)
                     .put(SharedBlobCacheWarmingService.ID_LOOKUP_PREWARM_RATIO_SETTING.getKey(), idLookupPreWarmRatio)
                     .build();
-            }
-
-            @Override
-            protected StatelessSharedBlobCacheService createCacheService(
-                NodeEnvironment nodeEnvironment,
-                Settings settings,
-                ThreadPool threadPool,
-                MeterRegistry meterRegistry
-            ) {
-                return spy(super.createCacheService(nodeEnvironment, settings, threadPool, meterRegistry));
-            }
-
-            @Override
-            public BlobContainer wrapBlobContainer(BlobPath path, BlobContainer innerContainer) {
-                return new FilterBlobContainer(innerContainer) {
-                    @Override
-                    protected BlobContainer wrapChild(BlobContainer child) {
-                        return child;
-                    }
-
-                    @Override
-                    public InputStream readBlob(OperationPurpose purpose, String blobName, long position, long length) throws IOException {
-                        return new InputStream() {
-                            private long remaining = length;
-
-                            @Override
-                            public int read() throws IOException {
-                                if (remaining == 0) {
-                                    return -1;
-                                } else {
-                                    remaining -= 1;
-                                    return 1;
-                                }
-                            }
-                        };
-                    }
-                };
             }
         };
     }
@@ -1271,6 +1229,58 @@ public class SharedBlobCacheWarmingServiceTests extends ESTestCase {
                     .build();
             }
         };
+    }
+
+    class FakeStatelessNodeWithFakeObjectStore extends FakeStatelessNode {
+        FakeStatelessNodeWithFakeObjectStore() throws IOException {
+            this(1);
+        }
+
+        FakeStatelessNodeWithFakeObjectStore(int primaryTerm) throws IOException {
+            super(
+                SharedBlobCacheWarmingServiceTests.this::newEnvironment,
+                SharedBlobCacheWarmingServiceTests.this::newNodeEnvironment,
+                xContentRegistry(),
+                primaryTerm
+            );
+        }
+
+        @Override
+        protected StatelessSharedBlobCacheService createCacheService(
+            NodeEnvironment nodeEnvironment,
+            Settings settings,
+            ThreadPool threadPool,
+            MeterRegistry meterRegistry
+        ) {
+            return spy(super.createCacheService(nodeEnvironment, settings, threadPool, meterRegistry));
+        }
+
+        @Override
+        public BlobContainer wrapBlobContainer(BlobPath path, BlobContainer innerContainer) {
+            return new FilterBlobContainer(innerContainer) {
+                @Override
+                protected BlobContainer wrapChild(BlobContainer child) {
+                    return child;
+                }
+
+                @Override
+                public InputStream readBlob(OperationPurpose purpose, String blobName, long position, long length) throws IOException {
+                    return new InputStream() {
+                        private long remaining = length;
+
+                        @Override
+                        public int read() throws IOException {
+                            if (remaining == 0) {
+                                return -1;
+                            } else {
+                                remaining -= 1;
+                                return 1;
+                            }
+                        }
+                    };
+                }
+            };
+        }
     }
 
     private void appendCommitsToVbcc(

